cpp/src/arrow/util/align_util.h

```diff
@@ -64,5 +66,23 @@ inline BitmapWordAlignParams BitmapWordAlign(const uint8_t* data, int64_t bit_of
   return p;
 }
 
+template <typename T>
+Status EnsureAlignment(std::shared_ptr<T> object, int64_t alignment,
+                       MemoryPool* memory_pool) {
+  std::vector<std::shared_ptr<Buffer>> buffers_ = object->data()->buffers;
+  for (size_t i = 0; i < buffers_.size(); ++i) {
+    if (buffers_[i]) {
+      auto buffer_address = buffers_[i]->address();
+      if ((buffer_address % alignment) != 0) {
+        ARROW_ASSIGN_OR_RAISE(
+            auto new_buffer, AllocateBuffer(buffers_[i]->size(), alignment, memory_pool));
+        std::memcpy(new_buffer->mutable_data(), buffers_[i]->data(), buffers_[i]->size());
+        object->data()->buffers[i] = std::move(new_buffer);
```

Comment:
It'd be nice (and more consistent with the rest of the code base) if we could do this by returning a new `std::shared_ptr<Array>` instead of modifying the existing array in-place.  You should be able to create a new `ArrayData` with the new buffers and then create a new `Array` from that.

Reply:
Made the change, thanks!